mate-panel/panel-widget.c

```diff
@@ -623,36 +626,92 @@ get_applet_list_pos (PanelWidget *panel,
 	return NULL;
 }
 
-/*tells us if an applet is "stuck" on the right side*/
-int
-panel_widget_is_applet_stuck (PanelWidget *panel_widget,
-			      GtkWidget   *widget)
+/*
+ * This function determines whether the given panel applet should be positioned
+ * relative to the center or end of the panel; that is, whether the applet
+ * is part of a contiguous group of other applets, one of which touches the
+ * center or end of the panel.
+ */
+PanelObjectEdgeRelativity
+panel_widget_determine_applet_edge_relativity (PanelWidget *panel_widget,
+					       GtkWidget   *widget)
 {
-	AppletData *applet;
+	AppletData *applet_in_question;
+	gboolean    applet_is_centered = FALSE;
+
+	g_return_val_if_fail (PANEL_IS_WIDGET (panel_widget), PANEL_EDGE_START);
+	g_return_val_if_fail (GTK_IS_WIDGET (widget), PANEL_EDGE_START);
```

Comment:
I don't understand this change. You're trying to determine whether the widget is actually a widget, but you're comparing against an enum. I suspect this change is a mistake.

Reply:
That's not what `g_return_val_if_fail ()` does.  It tests if its first parameter (say `PANEL_IS_WIDGET (panel_widget)`) is true (nonzero, in this case non-`NULL`), and if it is not true (i.e., it is `NULL`), then immediately terminate the current function, returning the value specified in the second argument to the function (actually `g_return_val_if_fail ()` is a pre-processor macro).

If you don't believe me, here's the official GLib documentation about the macro (in the code itself, at https://github.com/GNOME/glib/blob/main/glib/gmessages.h, lines 588-611 as I write this):

```
/**
 * g_return_val_if_fail:
 * @expr: the expression to check
 * @val: the value to return from the current function
 *       if the expression is not true
 *
 * Verifies that the expression @expr, usually representing a precondition,
 * evaluates to %TRUE. If the function does not return a value, use
 * g_return_if_fail() instead.
 *
 * If @expr evaluates to %FALSE, the current function should be considered to
 * have undefined behaviour (a programmer error). The only correct solution
 * to such an error is to change the module that is calling the current
 * function, so that it avoids this incorrect call.
 *
 * To make this undefined behaviour visible, if @expr evaluates to %FALSE,
 * the result is usually that a critical message is logged and @val is
 * returned from the current function.
 *
 * If `G_DISABLE_CHECKS` is defined then the check is not performed.  You
 * should therefore not depend on any side effects of @expr.
 *
 * See g_return_if_fail() for guidance on how to debug failure of this check.
 */
```

(By the way, the version of `g_return_val_if_fail ()` defined right below that comment is for if GLib was compiled with sanity checks disabled -- that version is a dummy macro that does nothing anyway; furhter below, at line 649, the real version used in most builds of GLib is defined:

```
#define g_return_val_if_fail(expr, val) \
  G_STMT_START { \
    if (G_LIKELY (expr)) \
      { } \
    else \
      { \
        g_return_if_fail_warning (G_LOG_DOMAIN, \
                                  G_STRFUNC, \
                                  #expr); \
        return (val); \
      } \
  } G_STMT_END
```

Not to be arrogant, but let's break that down:  The macro tests for the likely case that the first parameter [`expr`] evaluates to true or a nonzero / non-null value:  In that case, nothing happens [hence the empty `{}` brackets].  However, if that test fails and `expr` evaluates to false, zero or NULL, then the `else` block is run, a warning is printed on the terminal [or into `.xsession-errors`, more likely], and the second parameter [`val`] is returned.)

So, maybe you have a point, but I don't see what's wrong.  I'd appreciate to hear from you on this matter.  Thank you.